lib/SILGen/SILGenExpr.cpp

```diff
@@ -3021,24 +3021,23 @@ emitKeyPathRValueBase(SILGenFunction &subSGF,
                                              AbstractionPattern::getOpaque(),
                                              baseType);
   
-  // Upcast a class instance to the property's declared type if necessary.
-  if (auto propertyClass = dyn_cast<ClassDecl>(storage->getDeclContext())) {
-    if (baseType->getClassOrBoundGenericClass() != propertyClass) {
-      baseType = baseType->getSuperclassForDecl(propertyClass)
-        ->getCanonicalType();
-      paramSubstValue = subSGF.B.createUpcast(loc, paramSubstValue,
-                                     SILType::getPrimitiveObjectType(baseType));
+  // Pop open an existential container base.
+  if (baseType->isAnyExistentialType()) {
+    // Use the opened archetype from the AST for a generic member, or make a
+    // new one (which we'll upcast immediately below) for a class member.
+    ArchetypeType *opened;
+    if (storage->getDeclContext()->getSelfInterfaceType()
+                                 ->getClassOrBoundGenericClass()) {
+      opened = ArchetypeType::getOpened(baseType);
+    } else {
+      opened = subs[0].getReplacement()->castTo<ArchetypeType>();
```

Comment:
Is there a way to get at this without going through `subs[0]`? It's kind of a chicken and egg problem because we need to get rid of SubstitutionList but can't until the right abstractions are in place, but those require the representation not be a naked ArrayRef...

Reply:
The opened archetype also has to line up with the one the AST picked for the substitutions in the following property `ConcreteDeclRef`. If `ConcreteDeclRef` used a `SubstitutionMap` instead we could grab the substitution for `Self` here instead.